enzyme/Enzyme/AdjointGenerator.h

```diff
@@ -6654,10 +6654,115 @@ class AdjointGenerator
+      std::map<int, Type *> gradByVal;
+
+      for (unsigned i = 0; i < orig->getNumArgOperands(); ++i) {
+
+        auto argi = gutils->getNewFromOriginal(orig->getArgOperand(i));
+
+#if LLVM_VERSION_MAJOR >= 9
+        if (orig->isByValArgument(i)) {
+          gradByVal[args.size()] = orig->getParamByValType(i);
+        }
+#endif
+        args.push_back(argi);
+
+        if (gutils->isConstantValue(orig->getArgOperand(i)) &&
+            !foreignFunction) {
+          argsInverted.push_back(DIFFE_TYPE::CONSTANT);
+          continue;
+        }
+
+        auto argType = argi->getType();
+
+        if (!argType->isFPOrFPVectorTy() &&
+            (TR.query(orig->getArgOperand(i)).Inner0().isPossiblePointer() ||
+             foreignFunction)) {
+          DIFFE_TYPE ty = DIFFE_TYPE::DUP_ARG;
+          if (argType->isPointerTy()) {
+#if LLVM_VERSION_MAJOR >= 12
+            auto at = getUnderlyingObject(orig->getArgOperand(i), 100);
+#else
+            auto at = GetUnderlyingObject(
+                orig->getArgOperand(i),
+                gutils->oldFunc->getParent()->getDataLayout(), 100);
+#endif
+            if (auto arg = dyn_cast<Argument>(at)) {
+              if (constant_args[arg->getArgNo()] == DIFFE_TYPE::DUP_NONEED) {
+                ty = DIFFE_TYPE::DUP_NONEED;
+              }
+            }
+          }
+          args.push_back(
+              gutils->invertPointerM(orig->getArgOperand(i), Builder2));
+          argsInverted.push_back(ty);
+
+          // Note sometimes whattype mistakenly says something should be
+          // constant [because composed of integer pointers alone]
+          assert(whatType(argType, Mode) == DIFFE_TYPE::DUP_ARG ||
+                 whatType(argType, Mode) == DIFFE_TYPE::CONSTANT);
+        } else {
+          if (foreignFunction)
+            assert(!argType->isIntOrIntVectorTy());
+
+          args.push_back(diffe(orig->getArgOperand(i), Builder2));
+          argsInverted.push_back(DIFFE_TYPE::DUP_ARG);
+        }
+      }
+
+      auto newcalled = gutils->Logic.CreatePrimalAndGradient(
+          cast<Function>(called), subretType, argsInverted, gutils->TLI,
+          TR.analyzer.interprocedural, /*returnValue*/ retUsed,
+          /*subdretptr*/ false, DerivativeMode::ForwardMode, nullptr,
+          nextTypeInfo, uncacheable_args, nullptr,
+          /*AtomicAdd*/ gutils->AtomicAdd);
+
+      assert(newcalled);
+      FunctionType *FT = cast<FunctionType>(
+          cast<PointerType>(newcalled->getType())->getElementType());
+
+      CallInst *diffes = Builder2.CreateCall(FT, newcalled, args);
+      diffes->setCallingConv(orig->getCallingConv());
+      diffes->setDebugLoc(gutils->getNewFromOriginal(orig->getDebugLoc()));
+#if LLVM_VERSION_MAJOR >= 9
+      for (auto pair : gradByVal) {
+        diffes->addParamAttr(
+            pair.first,
+            Attribute::getWithByValType(diffes->getContext(), pair.second));
+      }
+#endif
+
+      if (!diffes->getType()->isVoidTy()) {
```

Comment:
Should this not be a check whether the value returned by the call is constant?

Reply:
yes!